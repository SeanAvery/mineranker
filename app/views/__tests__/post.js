import React from 'react'
import TestUtils from 'react-addons-test-utils'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { StyleSheetTestUtils } from 'aphrodite'

import PostView from './../post'
import Post from './../../components/post'
import Loader from './../../components/loader'

describe('<PostView {...props} />', () => {
  let defaultPost
  let defaultProps

  beforeEach(() => {
    StyleSheetTestUtils.suppressStyleInjection()

    defaultPost = {
      __v: 0,
      _id: 'test-id',
      comments: [],
      comments_count: 0,
      company_name: 'acme inc',
      created: 'Mon Jan 02 2012 00:00:00 GMT+0100 (CET)',
      likes: 42,
      photo_url: 'localhost/photo-url',
      user: {
        avatar_url: 'localhost/avatar-url',
        id: 'test-user-id',
        username: 'test-username',
      },
      video_url: 'localhost/video-url',
      voted: false,
    }

    defaultProps = {
      error: undefined,
      onLoadMore: () => {},
      onStartVideo: () => {},
      onStopVideo: () => {},
      loadMoreThreshold: 4,
      hasError: false,
      isWorking: false,
      posts: [defaultPost],
    }
  })

  afterEach(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection()
  })

  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<PostView {...defaultProps} />, div)
  })

  it('renders correctly', () => {
    const tree = renderer.create(
      <PostView {...defaultProps} posts={[{ ...defaultPost, video_url: null }]} />
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders multiple posts', () => {
    const posts = defaultProps.posts
    const newPost = {
      __v: 0,
      _id: 'test-id-1',
      comments: [],
      comments_count: 0,
      company_name: 'acme inc',
      created: 'Mon Jan 01 2012 00:00:00 GMT+0100 (CET)',
      likes: 87,
      photo_url: 'localhost/photo-url-1',
      user: {
        avatar_url: 'localhost/avatar-url',
        id: 'test-user-id',
        username: 'test-username',
      },
      video_url: 'localhost/video-url-1',
      voted: false,
    }
    const postViewPosts = [
      ...posts,
      newPost,
    ]

    const post = shallow(
      <PostView
        { ...defaultProps }
        posts={postViewPosts}
      />
    )

    const children = post.children().children(Post)
    expect(children.length).toEqual(2)
    children.forEach((node, i) => {
      expect(node.props()._id).toEqual(postViewPosts[i]._id)
    })
  })

  it('renders single post', () => {
    const post = shallow(
      <PostView
        { ...defaultProps }
      />
    )

    const children = post.children().children(Post)
    expect(children.length).toEqual(1)
  })

  it('renders zero posts', () => {
    const dom = (
      <PostView
        { ...defaultProps }
        posts={[]}
      />
    )
    const post = shallow(dom)

    const children = post.children().children(Post)
    expect(children.length).toEqual(0)

    // snapshot
    const tree = renderer.create(dom).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('loads more posts before hitting end of view', () => {
    const posts = defaultProps.posts
    const newPost = {
      __v: 0,
      _id: 'test-id-1',
      comments: [],
      comments_count: 0,
      company_name: 'acme inc',
      created: 'Mon Jan 01 2012 00:00:00 GMT+0100 (CET)',
      likes: 87,
      photo_url: 'localhost/photo-url-1',
      user: {
        avatar_url: 'localhost/avatar-url',
        id: 'test-user-id',
        username: 'test-username',
      },
      video_url: 'localhost/video-url-1',
      voted: false,
    }

    const mockOnLoadMore = jest.fn()

    const post = TestUtils.renderIntoDocument(
      <PostView
        { ...defaultProps }
        loadMoreThreshold={1}
        posts={[
          ...posts,
          newPost,
          newPost,
          newPost,
        ]}
        onLoadMore={() => mockOnLoadMore()}
      />
    )
    // Find the first post node and scroll
    const postNode = ReactDOM.findDOMNode(post)
    document.body.scrollTop = 500
    window.dispatchEvent(new window.UIEvent('scroll', { target: postNode, detail: 0 }))

    expect(mockOnLoadMore.mock.calls.length).toBe(1)
  })

  it('renders loading icon when loading more posts', () => {
    const dom = (
      <PostView
        { ...defaultProps }
        posts={[ { ...defaultPost, video_url: null } ]}
        isWorking={true}
      />
    )
    const post = shallow(dom)

    const children = post.children().children(Loader)
    expect(children.length).toBe(1)

    // snapshot
    const tree = renderer.create(dom).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('gracefully handles error if a problem was encountered', () => {
    const dom = (
      <PostView
        { ...defaultProps }
        hasError={true}
        posts={[]}
        error={'Problem connecting to the network'}
      />
    )
    const post = shallow(dom)

    const children = post.children().children('span')
    expect(children.length).toBe(1)

    // snapshot
    const tree = renderer.create(dom).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
