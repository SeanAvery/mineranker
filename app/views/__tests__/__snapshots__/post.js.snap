// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostView {...props} /> gracefully handles error if a problem was encountered 1`] = `
<div
  className="container_1hpj00t"
>
  <div
    className="header_dqzfq"
  >
    <img
      className="logo_gt6fx2"
      src="test-file-stub"
    />
    <a
      className="nav_1o07ka5-o_O-selected_104b05c"
      href="javascript:void(0)"
      onClick={[Function]}
    >
      news
    </a>
  </div>
  <div
    className="wrapper_1bc0z68"
  >
    <span
      className="error_13jehyk"
    >
      Problem connecting to the network
    </span>
  </div>
</div>
`;

exports[`<PostView {...props} /> renders correctly 1`] = `
<div
  className="container_1hpj00t"
>
  <div
    className="header_dqzfq"
  >
    <img
      className="logo_gt6fx2"
      src="test-file-stub"
    />
    <a
      className="nav_1o07ka5-o_O-selected_104b05c"
      href="javascript:void(0)"
      onClick={[Function]}
    >
      news
    </a>
  </div>
  <div
    className="wrapper_1bc0z68"
  >
    <div
      className="container_1ox283w"
    >
      <div
        className="container_1ssoezr"
      >
        <img
          className="image_kt1q8t"
          src={undefined}
        />
      </div>
      <div
        className="body_gexgkn"
      >
        <span
          className="title_njp6s6"
        />
        <span
          className="description_14rca9u"
        />
      </div>
      <div
        className="footer_diuw3w"
      >
        <div>
          <a
            className="vote_12qtyj8"
            href="javascript:void(0)"
            onClick={[Function]}
          >
            <img
              className="svg_1qn01rn"
              src="test-file-stub"
            />
            <span
              className="voteText_vxbm49"
            />
          </a>
          <a
            className="vote_12qtyj8"
            href="javascript:void(0)"
            onClick={[Function]}
          >
            <img
              className="svg_1qn01rn"
              src="test-file-stub"
            />
            <span
              className="voteText_vxbm49"
            />
          </a>
        </div>
        <div
          className="link_1e37oh4"
        >
          <a
            className="goto_12qtyj8"
            href={undefined}
            target="_blank"
          >
            <span
              className="gotoText_vxbm49"
            >
              go to site
            </span>
            <img
              className="svg_1qn01rn"
              src="test-file-stub"
            />
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PostView {...props} /> renders loading icon when loading more posts 1`] = `
<div
  className="container_1hpj00t"
>
  <div
    className="header_dqzfq"
  >
    <img
      className="logo_gt6fx2"
      src="test-file-stub"
    />
    <a
      className="nav_1o07ka5-o_O-selected_104b05c"
      href="javascript:void(0)"
      onClick={[Function]}
    >
      news
    </a>
  </div>
  <div
    className="wrapper_1bc0z68"
  >
    <div
      className="container_1ox283w"
    >
      <div
        className="container_1ssoezr"
      >
        <img
          className="image_kt1q8t"
          src={undefined}
        />
      </div>
      <div
        className="body_gexgkn"
      >
        <span
          className="title_njp6s6"
        />
        <span
          className="description_14rca9u"
        />
      </div>
      <div
        className="footer_diuw3w"
      >
        <div>
          <a
            className="vote_12qtyj8"
            href="javascript:void(0)"
            onClick={[Function]}
          >
            <img
              className="svg_1qn01rn"
              src="test-file-stub"
            />
            <span
              className="voteText_vxbm49"
            />
          </a>
          <a
            className="vote_12qtyj8"
            href="javascript:void(0)"
            onClick={[Function]}
          >
            <img
              className="svg_1qn01rn"
              src="test-file-stub"
            />
            <span
              className="voteText_vxbm49"
            />
          </a>
        </div>
        <div
          className="link_1e37oh4"
        >
          <a
            className="goto_12qtyj8"
            href={undefined}
            target="_blank"
          >
            <span
              className="gotoText_vxbm49"
            >
              go to site
            </span>
            <img
              className="svg_1qn01rn"
              src="test-file-stub"
            />
          </a>
        </div>
      </div>
    </div>
    <span
      className="fa fa-spinner fa-spin spinner_wdemd5-o_O-spinner_1tj8unv"
    />
  </div>
</div>
`;

exports[`<PostView {...props} /> renders zero posts 1`] = `
<div
  className="container_1hpj00t"
>
  <div
    className="header_dqzfq"
  >
    <img
      className="logo_gt6fx2"
      src="test-file-stub"
    />
    <a
      className="nav_1o07ka5-o_O-selected_104b05c"
      href="javascript:void(0)"
      onClick={[Function]}
    >
      news
    </a>
  </div>
  <div
    className="wrapper_1bc0z68"
  >
    <div
      className="noResults_1amdm9c"
    >
      Sorry, nothing found :(
    </div>
  </div>
</div>
`;
